File: matlab/lfmm3d_example.m
%

disp("This is an example matlab driver");
disp("On output the code prints the pot+grad in example1 and, pot+pottarg in example2");
disp(" ");
disp(" ");
%%   source to source, charges only, pot+grad example
%

clear srcinfo ns U eps pg;
disp("Example 1: source to source, charge, pot+grad");
disp(" ");
disp(" ");

ns = 40000;
srcinfo.sources = rand(3,ns);
srcinfo.charges = rand(1,ns);

eps = 1e-5;
pg = 2;
U = lfmm3d(eps,srcinfo,pg);
disp("pot=");
disp(U.pot(1:9));
disp("grad=");
disp(U.grad(:,1:3));

clear srcinfo ns U eps pg pgt;
%
%%    source to source+target, charges+dipoles, pot example
%
disp(" ");
disp(" ");
disp("Example 2: source to source+target, charge+dipole, multiple densities, pot");
disp(" ");
disp(" ");

ns = 4000;
nt = 3999;

nd = 5;
srcinfo.nd = nd;

pg = 1;
pgt = 1;

srcinfo.sources = rand(3,ns);
srcinfo.charges = rand(nd,ns);
srcinfo.dipoles = rand(nd,3,ns);

targ = rand(3,nt);
eps = 1e-5;

U = lfmm3d(eps,srcinfo,pg,targ,pgt);
disp("pot=");
disp(U.pot(:,1:3));
disp("pottarg=");
disp(U.pottarg(:,1:3));

clear srcinfo ns U eps pg pgt nt targ nd;

File: c/lfmm3d_example.c
#include "stdlib.h"
#include "stdio.h"
#include "math.h"
#include "lfmm3d_c.h"
#include "complex.h"
#include "cprini.h"

int main(int argc, char **argv)
{
  cprin_init("stdout", "fort.13");
  cprin_skipline(2);


  int ns=2000;
  int ier=0;
  double *source = (double *)malloc(3*ns*sizeof(double));

  double *charge = (double *)malloc(ns*sizeof(double));

  double *pot = (double *)malloc(ns*sizeof(double));
  double *grad = (double *)malloc(3*ns*sizeof(double));

// initialize arrays
  for(int i=0;i<ns;i++)
  {
    source[3*i] = pow(rand01(),2);
    source[3*i+1] = pow(rand01(),2);
    source[3*i+2] = pow(rand01(),2);

    charge[i] = rand01(); 

  }

  double eps = 0.5e-6;
  cprin_message("this code is an example c driver.");
  cprin_message("on output, the code prints sample pot,grad");
  cprin_skipline(2);

// call the fmm routine
  lfmm3d_s_c_g_(&eps, &ns, source, charge, pot, grad, &ier);


  cprind("pot=",pot,12);
  cprind("grad=",grad,12);

  return 0;
}
